posthog/models/sessions/sql.py

```diff
@@ -0,0 +1,131 @@
+    -- we want the fewest rows possible but also the fastest queries
+    -- since we query by date and not by time
+    -- and order by must be in order of increasing cardinality
+    -- so we order by date first, then team_id, then session_id
+    -- hopefully, this is a good balance between the two
+    ORDER BY (toDate(min_first_timestamp), team_id, session_id)
+SETTINGS index_granularity=512
+"""
+).format(
+    table_name=SESSIONS_DATA_TABLE(),
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=SESSIONS_DATA_TABLE_ENGINE(),
+)
+
+SESSIONS_TABLE_MV_SQL = (
+    lambda: """
+CREATE MATERIALIZED VIEW IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
+TO {database}.{target_table}
+AS SELECT
+$session_id as session_id,
+team_id,
+any(distinct_id) as distinct_id,
+min(timestamp) AS min_first_timestamp,
+max(timestamp) AS max_last_timestamp,
+-- TRICKY: ClickHouse will pick a relatively random "entry"
+-- when it collapses the aggregating merge tree
+-- unless we teach it what we want...
+-- argMin ignores null values
+-- so this will get the first non-null value of "entry"
+-- for each group of session_id and team_id
+-- by min of first_timestamp in the batch
+-- this is an aggregate function, not a simple aggregate function
+-- so we have to write to argMinState, and query with argMinMerge
+
+groupUniqArray(JSONExtractString(properties, '$current_url')) AS urls,
+argMinState(JSONExtractString(properties, '$current_url'), timestamp) as entry_url,
+argMaxState(JSONExtractString(properties, '$current_url'), timestamp) as exit_url,
+argMinState(JSONExtractString(properties, '$initial_utm_source'), timestamp) as initial_utm_source,
+argMinState(JSONExtractString(properties, '$initial_utm_campaign'), timestamp) as initial_utm_campaign,
+argMinState(JSONExtractString(properties, '$initial_utm_medium'), timestamp) as initial_utm_medium,
+argMinState(JSONExtractString(properties, '$initial_utm_term'), timestamp) as initial_utm_term,
+argMinState(JSONExtractString(properties, '$initial_utm_content'), timestamp) as initial_utm_content,
+argMinState(JSONExtractString(properties, '$initial_referring_domain'), timestamp) as initial_referring_domain,
+count(*) as event_count,
+sumIf(1, event='$pageview') as pageview_count
+
+FROM {database}.events
+group by session_id, team_id
+""".format(
+        table_name=f"{TABLE_BASE_NAME}_mv",
+        target_table=f"writable_{TABLE_BASE_NAME}",
+        cluster=settings.CLICKHOUSE_CLUSTER,
+        database=settings.CLICKHOUSE_DATABASE,
+    )
+)
+
+# Distributed engine tables are only created if CLICKHOUSE_REPLICATED
+
+# This table is responsible for writing to sharded_sessions based on a sharding key.
+WRITABLE_SESSIONS_TABLE_SQL = lambda: SESSIONS_TABLE_BASE_SQL.format(
+    table_name=f"writable_{TABLE_BASE_NAME}",
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=Distributed(
+        data_table=SESSIONS_DATA_TABLE(),
+        sharding_key="sipHash64(distinct_id)",
```

Comment:
possibly we should shard by session id... (although it'll be often functionally the same)

Reply:
(I actually don't know how to decide 🤯)

Reply:
https://posthog.com/handbook/engineering/clickhouse/schema/sharded-events#sharding-by-cityhash64distinct_id

I wondering whether sharding by session_id would be better, you'll always be grouping by session_id so you want CH to know it doesn't have to copy the data over the network to the coordinator node because all of the data will be on one shard

But 🤷

Reply:
Agreed, I think that's probably better too for querying